### PR TITLE
Web: Add support for client search query params for Table (have to opt in)

### PR DESCRIPTION
**Pull request by kimlisa**
required by https://github.com/gravitational/teleport.e/pull/4784

enterprise changes: https://github.com/gravitational/teleport.e/pull/4984

recommend reviewing by commit

- Add support for client search query params for Table
- Searching now requires user to press `enter` to perform the action. This is now consistent with other custom search bars (eg. unified resource, access list, etc)
- refactored encodeUrlQueryParams props to accept object props
- random: Allow passing flex basis as a prop to HoverTooltip.tsx

### CODE REVIEW

**Comment by ryanclark** [comment]
> The search query param does not get updated as user types on the search bar

Why not?

**Comment by kimlisa** [comment]
> > The search query param does not get updated as user types on the search bar
> 
> Why not?

it's supported now, i initially didn't want to get blocked in case okta stuff needed to go in first, but it ended up not being ready anyways

File: web/packages/design/src/DataTable/types.ts [hunk 1]
```diff
@@ -35,6 +35,22 @@ export type TableProps<T> = {
    */
   emptyHint?: string;
   pagination?: PaginationConfig<T>;
+  /**
+   * config for client searching.
+   * supports any table except when "serversideProps"
+   * field is defined
+   */
+  clientSearchConfig?: {
```
Comment (ravicious) [inline web/packages/design/src/DataTable/types.ts:43]:
```suggestion
  clientSearch?: {
```

File: web/packages/design/src/DataTable/types.ts [hunk 2]
```diff
@@ -35,6 +35,22 @@ export type TableProps<T> = {
...
+    /**
+     * By default, no initial search is applied (empty search),
+     * unless "initialSearchValue" is defined.
+     */
+    initialSearchValue: string;
+    /**
+     * After setting a new search value, this function will be called.
+     */
+    updateUrlQueryParams(searchString: string): void;
```
Comment (ravicious) [inline web/packages/design/src/DataTable/types.ts:52]:
```suggestion
    onSearchValueChange(searchValue string): void;
```

File: web/packages/design/src/DataTable/useTable.ts [hunk 3]
```diff
@@ -50,8 +51,8 @@ export default function useTable<T>({
     }
 
     return {
-      data: serversideProps || disableFilter ? data : [],
```
Comment (ravicious) [inline web/packages/design/src/DataTable/useTable.ts:53]: Why are we getting rid of this behavior? It seems unrelated to client-side search.

File: web/packages/design/src/DataTable/useTable.ts [hunk 4]
```diff
@@ -61,7 +62,7 @@ export default function useTable<T>({
         : null,
       pagination: pagination
         ? {
-            paginatedData: paginateData(data, pagination.pageSize),
```
Comment (ravicious) [inline web/packages/design/src/DataTable/useTable.ts:64]: Same here.

File: web/packages/teleport/src/components/ClientSearcheableTableWithQueryParamSupport/ClientSearcheableTableWithQueryParamSupport.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+import { TableProps } from 'design/DataTable/types';
+import Table from 'design/DataTable';
+
+import {
+  decodeUrlQueryParam,
+  encodeUrlQueryParams,
+} from '../hooks/useUrlFiltering';
+
+export function ClientSearcheableTableWithQueryParamSupport<T>(
```
Comment (ravicious) [inline web/packages/teleport/src/components/ClientSearcheableTableWithQueryParamSupport/ClientSearcheableTableWithQueryParamSupport.tsx:29]: Is there a client-searchable table without query param support? If not then I think we might as well drop `WithQueryParamSupport` from the name. ;)

If someone wants to have a client-searchable table that doesn't update the query params, then they can just use a plain `<DataTable>`

File: web/packages/teleport/src/components/ClientSearcheableTableWithQueryParamSupport/ClientSearcheableTableWithQueryParamSupport.tsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  props: TableProps<T>
+) {
+  const searchParams = new URLSearchParams(location.search);
```
Comment (ravicious) [inline web/packages/teleport/src/components/ClientSearcheableTableWithQueryParamSupport/ClientSearcheableTableWithQueryParamSupport.tsx:32]: Should we use `useLocation` from react-router instead? It looks like we do use it in quite a few places already.

File: web/packages/teleport/src/components/ClientSearcheableTableWithQueryParamSupport/ClientSearcheableTableWithQueryParamSupport.tsx (hunk 5 above)
Comment (ravicious) [reply web/packages/teleport/src/components/ClientSearcheableTableWithQueryParamSupport/ClientSearcheableTableWithQueryParamSupport.tsx:29]: BTW, does it mean that client-side search is always enabled unless I specify `serversideProps`?

File: web/packages/design/src/DataTable/InputSearch/InputSearch.tsx [hunk 7]
```diff
@@ -34,22 +34,30 @@ export default function InputSearch({
   children,
   bigInputSize = false,
 }: Props) {
+  function submitSearch(e: FormEvent<HTMLFormElement>) {
+    e.preventDefault(); // prevent form default
+    const { searchValue } = e.target as typeof e.target & {
+      searchValue: { value: string };
+    };
+
+    setSearchValue(searchValue.value);
```
Comment (ravicious) [inline web/packages/design/src/DataTable/InputSearch/InputSearch.tsx:43]: I think there's too much stuff happening and we obscure too much how this works.

`e.target` points to a `HTMLFormElement` of the form. `e.target.searchValue` points to the search input DOM node because [that's how forms in DOM work](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement):

> Named inputs are added to their owner form instance as properties, and can overwrite native properties if they share the same name (e.g. a form with an input named action will have its action property return that input instead of the form's action HTML attribute).

And then `searchValue.value` reads that data from the DOM node.

To add to that, casting `e.target` to a specific "made-up" type is not necessarily wrong, but we don't have to do that. `searchValue` exists on `e.target` only because of that implicit browser behavior described above. In such cases, we can get away with doing `e.target["searchValue"]` instead. It gives us the same type-level guarantees as saying `e.target as typeof e.target & { searchValue: …`, which is just us saying "Trust me, `searchValue` is really there on this object I swear 🤞".

---

What we could do instead is use `FormData` and through this also signal that we're depending on some implicit browser behavior.

```diff
diff --git a/web/packages/design/src/DataTable/InputSearch/InputSearch.tsx b/web/packages/design/src/DataTable/InputSearch/InputSearch.tsx
index 817ba47369..d4f15b271d 100644
--- a/web/packages/design/src/DataTable/InputSearch/InputSearch.tsx
+++ b/web/packages/design/src/DataTable/InputSearch/InputSearch.tsx
@@ -28,6 +28,8 @@ import {
   ColorProps,
 } from 'design/system';
 
+const searchInputName = 'searchValue';
+
 export default function InputSearch({
   searchValue,
   setSearchValue,
@@ -36,11 +38,10 @@ export default function InputSearch({
 }: Props) {
   function submitSearch(e: FormEvent<HTMLFormElement>) {
     e.preventDefault(); // prevent form default
-    const { searchValue } = e.target as typeof e.target & {
-      searchValue: { value: string };
-    };
+    const formData = new FormData(e.currentTarget);
+    const searchValue = formData.get(searchInputName) as string;
 
-    setSearchValue(searchValue.value);
+    setSearchValue(searchValue);
   }
 
   return (
@@ -51,7 +52,7 @@ export default function InputSearch({
           placeholder="Search..."
           px={3}
           defaultValue={searchValue}
-          name="searchValue"
+          name={searchInputName}
           autoFocus
         />
         <ChildWrapperBackground>

```

Admittedly, we still have to do `formData.get(searchInputName) as string`, because technically `FormData.prototype.get` can return either a string or a file (if input's type was set to `file`). But we clearly show that we depend on the values from the DOM and through the `searchInputName` we make sure that we always read data from the correct input.

File: web/packages/teleport/src/config.ts [hunk 8]
```diff
@@ -557,8 +557,7 @@ const cfg = {
   },
 
   getUsersRoute() {
-    const clusterId = cfg.proxyCluster;
-    return generatePath(cfg.routes.users, { clusterId });
```
Comment (ravicious) [inline web/packages/teleport/src/config.ts:561]: Why is `clusterId` being removed from here?

File: web/packages/design/src/DataTable/useTable.ts (hunk 3 above)
Comment (kimlisa) [reply web/packages/design/src/DataTable/useTable.ts:53]: we have a [useEffect](https://github.com/gravitational/teleport/blob/564c245ecae10187371c1f8ab056093104ef5f31/web/packages/design/src/DataTable/useTable.ts#L192) that will set the `data`

before, it caused a brief flashing where the full list showed, until the searchValue got applied

File: web/packages/teleport/src/components/ClientSearcheableTableWithQueryParamSupport/ClientSearcheableTableWithQueryParamSupport.tsx (hunk 5 above)
Comment (kimlisa) [reply web/packages/teleport/src/components/ClientSearcheableTableWithQueryParamSupport/ClientSearcheableTableWithQueryParamSupport.tsx:29]: > Is there a client-searchable table without query param support?

yes, the query param support is opt in (you have to supply the `tableProps.clientSearch`) to make it work. i figured we can convert the rest of table by "need" basis

> BTW, does it mean that client-side search is always enabled unless I specify serversideProps?

if you use this table yes. if you supply `serversideProps`, the `clientSearch` gets ignored. the table that uses `serversideProps` provides us their own `serversideSearchPanel` component to handle its own server side searching.

I updated this tables props to disallow `serverSideProps`

File: web/packages/teleport/src/config.ts (hunk 8 above)
Comment (kimlisa) [reply web/packages/teleport/src/config.ts:561]: the user [route](https://github.com/gravitational/teleport/blob/17b395ecaa0dabd70781084d35e7081f4335d48d/web/packages/teleport/src/config.ts#L161) never needed it 😅 (or perhaps some left over from long ago?)

File: web/packages/teleport/src/Users/UserList/UserList.tsx
Comment (ravicious) [inline web/packages/teleport/src/Users/UserList/UserList.tsx:1]: Probably unrelated to changes within this PR, but when I search for `o`, the list of users doesn't get filtered at all and I see users who don't have `o` anywhere in their usernames or roles. I think it's because we search through the "Type" column as well.

**Review by ravicious — APPROVED** [review]
Looks good, though there's one failing test.

File: web/packages/teleport/src/Users/UserList/UserList.tsx
Comment (kimlisa) [reply web/packages/teleport/src/Users/UserList/UserList.tsx:1]: yeah we allow filtering by `type` as well, do you think it's not useful? 🤔

Comment (ravicious) [reply web/packages/teleport/src/Users/UserList/UserList.tsx:1]: I expect the search to work for user-provided values, so just the username and the roles, but it's just me.

**Review by ryanclark — APPROVED** [review]
